mm2src/coins/eth/eth_swap_v2/eth_maker_swap_v2.rs

```diff
@@ -0,0 +1,509 @@
+use crate::coin_errors::{ValidatePaymentError, ValidatePaymentResult};
+use crate::eth::{decode_contract_call, get_function_input_data, wei_from_big_decimal, EthCoin, EthCoinType,
+                 MakerPaymentStateV2, SignedEthTx, MAKER_SWAP_V2};
+use crate::{ParseCoinAssocTypes, RefundMakerPaymentSecretArgs, RefundMakerPaymentTimelockArgs, SendMakerPaymentArgs,
+            SpendMakerPaymentArgs, SwapTxTypeWithSecretHash, TransactionErr, ValidateMakerPaymentArgs,
+            WaitForPaymentSpendError};
+use common::executor::Timer;
+use common::now_sec;
+use ethabi::{Function, Token};
+use ethcore_transaction::Action;
+use ethereum_types::{Address, Public, U256};
+use ethkey::public_to_address;
+use futures::compat::Future01CompatExt;
+use mm2_err_handle::mm_error::MmError;
+use mm2_err_handle::prelude::{MapToMmResult, MmResult};
+use std::convert::TryInto;
+use web3::types::TransactionId;
+
+const ETH_MAKER_PAYMENT: &str = "ethMakerPayment";
+const ERC20_MAKER_PAYMENT: &str = "erc20MakerPayment";
+
+struct MakerPaymentArgs {
+    taker_address: Address,
+    taker_secret_hash: [u8; 32],
+    maker_secret_hash: [u8; 32],
+    payment_time_lock: u64,
+}
+
+struct MakerValidationArgs<'a> {
+    swap_id: Vec<u8>,
+    amount: U256,
+    taker: Address,
+    taker_secret_hash: &'a [u8],
+    maker_secret_hash: &'a [u8],
+    payment_time_lock: u64,
+}
+
+struct MakerRefundArgs {
+    payment_amount: U256,
+    taker_address: Address,
+    taker_secret: [u8; 32],
+    taker_secret_hash: [u8; 32],
+    maker_secret_hash: [u8; 32],
+    payment_time_lock: u64,
+    token_address: Address,
+}
+
+impl EthCoin {
+    pub(crate) async fn send_maker_payment_v2_impl(
+        &self,
+        args: SendMakerPaymentArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let maker_swap_v2_contract = self
+            .swap_v2_contracts
+            .as_ref()
+            .map(|contracts| contracts.maker_swap_v2_contract)
+            .ok_or_else(|| TransactionErr::Plain(ERRL!("Expected swap_v2_contracts to be Some, but found None")))?;
+        let payment_amount = try_tx_s!(wei_from_big_decimal(&args.amount, self.decimals));
+        let payment_args = {
+            let taker_address = public_to_address(args.taker_pub);
+            MakerPaymentArgs {
+                taker_address,
+                taker_secret_hash: try_tx_s!(args.taker_secret_hash.try_into()),
+                maker_secret_hash: try_tx_s!(args.maker_secret_hash.try_into()),
+                payment_time_lock: args.time_lock,
+            }
+        };
+        match &self.coin_type {
+            EthCoinType::Eth => {
+                let data = try_tx_s!(self.prepare_maker_eth_payment_data(&payment_args).await);
+                self.sign_and_send_transaction(
+                    payment_amount,
+                    Action::Call(maker_swap_v2_contract),
+                    data,
+                    U256::from(self.gas_limit_v2.maker.eth_payment),
```

Comment:
Can't we use `EthPaymentType` and `PaymentMethod` to get gas limits from the `gas_limit` function and abstract this away?

Reply:
> So abstraction here is overcomplicating code, its better to manage Eth and Erc20 separately

It will involve only adding one line as below
```rust
        let (_, gas_limit) = self
            .gas_limit_v2
            .gas_limit(&self.coin_type, EthPaymentType::MakerPayments, PaymentMethod::Send)
            .map_err(|e| TransactionErr::Plain(ERRL!("{}", e)))?;
        match &self.coin_type {
            EthCoinType::Eth => {
                let data = try_tx_s!(self.prepare_maker_eth_payment_data(&payment_args).await);
                self.sign_and_send_transaction(
                    payment_amount,
                    Action::Call(maker_swap_v2_contract),
                    data,
                    U256::from(gas_limit),
                )
                .compat()
                .await
            },
            EthCoinType::Erc20 {
                platform: _,
                token_addr,
            } => {
                let data = try_tx_s!(
                    self.prepare_maker_erc20_payment_data(&payment_args, payment_amount, *token_addr)
                        .await
                );
                self.handle_allowance(maker_swap_v2_contract, payment_amount, args.time_lock)
                    .await?;
                self.sign_and_send_transaction(
                    U256::from(ZERO_VALUE),
                    Action::Call(maker_swap_v2_contract),
                    data,
                    U256::from(gas_limit),
                )
                .compat()
                .await
            },
            EthCoinType::Nft { .. } => Err(TransactionErr::ProtocolNotSupported(ERRL!(
                "NFT protocol is not supported for ETH and ERC20 Swaps"
            ))),
        }
```
The benefit will be that we only access those fields from one place. We can then make these fields non-public. But I will leave this up to you, please resolve this if you still think it's unneeded.

Reply:
> It will involve only adding one line as below

Oh, I see, just decided you wanted to avoid having `match &self.coin_type` in send payment function and abstract everything

> The benefit will be that we only access those fields from one place. We can then make these fields non-public. But I will leave this up to you, please resolve this if you still think it's unneeded.

Hmm there is also `approve_payment` variant for taker. It is common for Eth and Erc20 (smart contract `takerPaymentApprove` method doesnt split logic for eth and erc20, so gas usage is the same)

https://github.com/KomodoPlatform/komodo-defi-framework/blob/72ed3060ace61d9a14eca6fc5cedeb957ad0be54/mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs#L220-L225

I suggest to keep `pub(crate)` visibility for `gas_limit_v2` for now, as there are still some places where its preferable to reach gas limit field directly.

I will leave this conversation open, to keep it noticeable for other people (and for us if we decide to return to it in future)